### PR TITLE
feat: read rows query model class

### PR DESCRIPTION
**Pull request by daniel-sanche**
Implements the RowQuery class, used to configure read_rows calls

Implementation is generally aligned with the [Query class in the Java library](https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/Query.java)

Implementation for sharding will come in a future PR

---

Note: This is merging into a new v3 branch, not main. I plan on making all PRs related to this rewrite to v3 and then merging v3 into main when development is complete

### CODE REVIEW

File: tests/unit/test_read_rows_query.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        query = self._make_one(["row_key_1", "row_key_2"], limit=10, row_filter=filter_)
+        self.assertEqual(len(query.row_keys), 2)
+        self.assertIn("row_key_1".encode(), query.row_keys)
+        self.assertIn("row_key_2".encode(), query.row_keys)
+        self.assertEqual(query.row_ranges, [])
+        self.assertEqual(query.filter, filter_)
+        self.assertEqual(query.limit, 10)
+
+    def test_ctor_invalid_limit(self):
+        with self.assertRaises(ValueError):
```
Comment (Mariatta) [inline tests/unit/test_read_rows_query.py:53]: Can you also assert that the valueerror message == "limit must be >= 0"?

File: tests/unit/test_read_rows_query.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        self.assertEqual(query.row_ranges[3][1].key, b"test_row7")
+        self.assertEqual(query.row_ranges[3][0].is_inclusive, False)
+        self.assertEqual(query.row_ranges[3][1].is_inclusive, True)
+        # test with nothing passed
+        result = query.add_range()
+        self.assertEqual(len(query.row_ranges), 5)
+        self.assertEqual(query.row_ranges[4][0], None)
+        self.assertEqual(query.row_ranges[4][1], None)
+        # test with inclusive flags only
+        with self.assertRaises(ValueError):
```
Comment (Mariatta) [inline tests/unit/test_read_rows_query.py:215]: `add_range` seems to be raising ValueError with varying messages. Perhaps we can also check which error message is being raised in each of the errors below.

Comment (daniel-sanche) [reply tests/unit/test_read_rows_query.py:215]: good point, I added that to all the raises statements

File: google/cloud/bigtable/read_rows_query.py [hunk 3]
```diff
@@ -13,36 +13,158 @@
...
     """
     Class to encapsulate details of a read row request
     """
 
     def __init__(
-        self, row_keys: list[str | bytes] | str | bytes | None = None, limit=None
+        self,
+        row_keys: list[str | bytes] | str | bytes | None = None,
+        limit: int | None = None,
+        row_filter: RowFilter | dict[str, Any] | None = None,
```
Comment (mutianf) [inline google/cloud/bigtable/read_rows_query.py:41]: __init__ doesn't have row_ranges. is this intended?

File: google/cloud/bigtable/read_rows_query.py [hunk 4]
```diff
@@ -13,36 +13,158 @@
...
+          - start_is_inclusive: if True, the start key is included in the range
+              defaults to True if None. Must not be included if start_key is None
+          - end_is_inclusive: if True, the end key is included in the range
+              defaults to False if None. Must not be included if end_key is None
+        """
+        # check for invalid combinations of arguments
+        if start_is_inclusive is None:
+            start_is_inclusive = True
+        elif start_key is None:
+            raise ValueError("start_is_inclusive must not be set without start_key")
```
Comment (mutianf) [inline google/cloud/bigtable/read_rows_query.py:145]: Maybe rephrase it to 
```suggestion
            raise ValueError("start_is_inclusive must be set with start_key")
```
so it's easier to understand :)

File: google/cloud/bigtable/read_rows_query.py (hunk 3 above)
Comment (daniel-sanche) [reply google/cloud/bigtable/read_rows_query.py:41]: Yes, because this implementation doesn't have an explicit range object, so passing in multiple ranges at init time would be difficult. Users can call `Query().add_range(start, end)` to add ranges

I think Python's style is biased towards flattness, and avoiding custom objects when you can get away with it. But I'm not opposed to adding a range class if you expect users to be re-using or passing ranges around a lot

File: google/cloud/bigtable/read_rows_query.py [hunk 5]
```diff
@@ -13,36 +13,158 @@
...
+from google.cloud.bigtable.row_filters import RowFilter
 
 if TYPE_CHECKING:
-    from google.cloud.bigtable.row_filters import RowFilter
     from google.cloud.bigtable import RowKeySamples
 
 
+@dataclass
+class _RangePoint:
+    # model class for a point in a row range
```
Comment (Mariatta) [inline google/cloud/bigtable/read_rows_query.py:27]:
```suggestion
    """Model class for a point in a row range"""
```

**Review by Mariatta — COMMENTED** [review]
One small change, but otherwise LGTM

**Review by Mariatta — APPROVED** [review]
